src/reaktoro_pse/tutorials/softening_acid_ro_example.ipynb

```diff
@@ -0,0 +1,1321 @@
+    "solver = get_cyipopt_watertap_solver(max_iter=200)\n",
+    "\n",
+    "result = solver.solve(m, tee=True)\n",
+    "assert_optimal_termination(result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 7. Add costing objective and setup optimizations\n",
+    "\n",
+    "1. HCl cost = 0.1 $/kg\n",
+    "2. CaO cost = 0.25 $/kg\n",
+    "3. N2CO3 cost = 0.4 $/kg\n",
+    "4. Constrain process to maintain Scaling Tendency to be less then 1\n",
+    "\n",
+    "Model will optimize the dosing of the 3 chemicals to minimize their cost, while ensuring not scaling occurs."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "m.fs.chemical_costs=Param(['HCl','CaO','Na2CO3'],initialize=1,units=pyunits.dimensionless)\n",
+    "m.fs.chemical_costs['HCl'].set_value(0.1)\n",
+    "m.fs.chemical_costs['CaO'].set_value(0.25)\n",
+    "m.fs.chemical_costs['Na2CO3'].set_value(0.4)\n",
+    "\n",
+    "m.fs.normalized_chemical_costs = Expression(\n",
+    "    expr=(\n",
+    "        m.fs.chemical_costs[\"CaO\"]\n",
+    "        * m.fs.softening.reagents_mol_flow[\"CaO\"]\n",
+    "        * m.fs.softening.reagents_mw[\"CaO\"]\n",
+    "        + m.fs.chemical_costs[\"Na2CO3\"]\n",
+    "        * m.fs.softening.reagents_mol_flow[\"Na2CO3\"]\n",
+    "        * m.fs.softening.reagents_mw[\"Na2CO3\"]\n",
+    "        + m.fs.chemical_costs[\"HCl\"]\n",
+    "        * m.fs.acid_addition.reagents_mol_flow[\"HCl\"]\n",
+    "        * m.fs.acid_addition.reagents_mw[\"HCl\"]\n",
+    "    )\n",
+    "    / (m.fs.desalter.water_removed * m.fs.feed.specie_mw['H2O'] / m.fs.feed.density)\n",
+    ")  # $/m3\n",
+    "\n",
+    "m.fs.cost_objective = Objective(expr=m.fs.normalized_chemical_costs)\n",
+    "\n",
+    "\n",
+    "m.fs.desalter.scaling_tendency[\"Calcite\"].setub(1)\n",
+    "m.fs.desalter.scaling_tendency[\"Gypsum\"].setub(1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "##### 7.1 Unfix chemical mol flows, so they can be optimized to satisfy our scaling constraints and minimize cost"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Unfix chemicals\n",
+    "m.fs.softening.reagents_mol_flow[\"CaO\"].unfix()\n",
+    "m.fs.softening.reagents_mol_flow[\"Na2CO3\"].unfix()\n",
+    "m.fs.acid_addition.reagents_mol_flow[\"HCl\"].unfix()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 8. Run simple analysis against water recovery"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Configure runs over differnt water recoveries. \n",
+    "\n",
+    "This can take a while as each solve takes 10-15 seconds. In full implementations uses ReaktoroBlockManager to run Reaktoro block solvers in parallel to reduce computational time. "
```

Comment:
In full implementations, we can use ReaktoroBlockManager to run Reaktoro block solvers in parallel to reduce computational time.